### ISSUE TITLE
Annotation's color

### ISSUE THREAD

**Issue by voodoorai2000** [consul/consul#1585]
The color of highlighted text should be according to a score type: 
1 point for each comment in that word, 1 point for each positive vote for each comment in that word
![color highlight legislation](https://cloud.githubusercontent.com/assets/4169/26720367/3f888c7a-4788-11e7-9696-7bbf5cc4c9e5.png)

Please send PR against the [medialab-legislation branch](https://github.com/consul/consul/compare/medialab-legislation?expand=1)

**Comment by bertocq** [comment]
@voodoorai2000 I've been trying to come up with a solution for this, but my CSS-fu or googling skills aren't enough to just determine if its 100% possible to have a simple solution or we've to go for the complex one 😞 .

## Current status:
- 👍  Each annotation has its own `.weight` property, and that's used to choose one [`weight-*` style](https://github.com/consul/consul/blob/master/app/assets/stylesheets/annotator_overrides.scss#L106-L120)

- **Scenario 1 Nested Annotations**: Outer annotation `Minima aut totam voluptatem` with 3 comments, and inner one `aut totam` with 1 comment) it looks like:
![screen shot 2017-06-11 at 22 41 07](https://user-images.githubusercontent.com/983242/27014327-1f4c9e9e-4ef7-11e7-96b5-ddba37887921.jpg)❌ 
Instead of :
![screen shot 2017-06-11 at 22 40 44](https://user-images.githubusercontent.com/983242/27014326-1f389944-4ef7-11e7-907e-3210f7dfe415.jpg)✅
Because the inner `aut totam` doesn't know it's inside another annotation with weight 3, it just knows about its own weight 😔 . 
Html is:
```html
<span class="annotator-hl weight-3" data-annotation-id="10">
   Minima
   <span class="annotator-hl weight-1" data-annotation-id="11">aut totam</span> 
   voluptatem
</span>
```

- **Scenario 2 Overlapping Annotations**: Given the text `dolorem odio laboriosam` with one annotation `dolorem odio` with 1 comment, and second annotation `odio laboriosam` with 1 comment.. the result looks like:
![screen shot 2017-06-11 at 22 13 32](https://user-images.githubusercontent.com/983242/27014164-6ec2641c-4ef3-11e7-896d-4dfb6fc67192.jpg)❌ 
But should be:
![screen shot 2017-06-11 at 22 13 49](https://user-images.githubusercontent.com/983242/27014165-7146f4b4-4ef3-11e7-92a6-02045d19cec6.jpg)✅ 
Html is:
```html
<span class="annotator-hl weight-2" data-annotation-id="8">dolorem </span>
<span class="annotator-hl weight-2" data-annotation-id="9">
   <span class="annotator-hl weight-2" data-annotation-id="8">odio</span>
   laboriosam
</span>
```

## My initial approach
CSS Background inheritance 💃  ?? Seems like that doesn't exist in CSS standard... and I couldn't imagine any kind of "hack" to create it... (I've tried using opacity but that messes up the text opacity 😢)  **maybe @decabeza you know some advance trick?**

The idea was to just to make any `span.weight-1` inside a `span.weight-3` look like if the class was `weight-4` somehow making the background colours overlap/inherit... but I think I'm asking for 🚀  science 😂 .

## The solution I'm heading towards
The root problem is that a `Legislation::Annotation` is not aware of other annotations in the same text. But any idea I think about its nuts:

**A**: ~Using graph-alike relationships (one annotation text can be as well included in two bigger ones, and enclose another 2 (and also the first includes the second)... could be a solution, but looks too messy for me from the data structure point of view, as well from the implementation (adding one node to the graph, and removing them as well)~ 👎 

**B**: ~Going to the word-level could be another option, but since words repeat, it would be impossible to calculate the weight of each word just going through all text annotations to see if it's included.~ 👎 

**C**: After setting all `weight-*` classes, loop again through all annotations and for each one add the weight of any `span` that includes it. 

Maybe this is the only reasonable-to-implement one, although my JS is not very good to write this down in a few ^_^ (that's mostly why I was going nuts about backend-ruby-implemented solutions).

**Comment by voodoorai2000** [comment]
Excellent @bertocq 

In my opinion, the eventual perfect solution is to make annotations aware of other annotations in the same range. But that is not trivial.

For now +1 to option C 👌

**Comment by xuanxu** [comment]
@bertocq To get a darker color on each overlaping annotation, instead of setting a weight, you can set opacity/transparency to the background color, that way it accumulates.

**Comment by bertocq** [comment]
@xuanxu FUUUUUUUUUUUUU You're right!! 👏 👏  I looked just at `opacity` css attribute as I mentioned, but that affects text as well so I discarded after some tries. But just right now I found out  rgba lets you set opacity for a color https://stackoverflow.com/a/5135033
```
rgba(255,0,0,0.1)    /* 10% opaque red */  
rgba(255,0,0,0.4)    /* 40% opaque red */  
rgba(255,0,0,0.7)    /* 70% opaque red */  
rgba(255,0,0,  1)    /* full opaque red */ 
```
I'm so used to hex values for colours that totally forgot that rgba, hsla, etc... exists ^_^. 

So:
![screen shot 2017-06-12 at 12 42 02](https://user-images.githubusercontent.com/983242/27030446-9187187a-4f6c-11e7-8fc5-eb20839fed89.jpg)

I'm going to do some tests to see if just those changes would do the trick and PR afterwards

**Comment by xuanxu** [comment]
@bertocq 👌

**Comment by bertocq** [comment]
### Using #FFE95F rgba with 0.2 to 1 opacities
Its really hard to tell the difference between 0.4+0.2 opacity and a 0.4... on the first scenario:
![screen shot 2017-06-12 at 12 56 43](https://user-images.githubusercontent.com/983242/27030976-a7e90d1a-4f6e-11e7-89e2-fa7a837f5c79.jpg)
The second one is a bit more clear:
![screen shot 2017-06-12 at 12 57 33](https://user-images.githubusercontent.com/983242/27030999-bdb3ac4a-4f6e-11e7-99c4-a65c3d435733.jpg)

### Using original hex colours with 0.2 to 1 opacities
![screen shot 2017-06-12 at 13 06 06](https://user-images.githubusercontent.com/983242/27031266-ea46cf98-4f6f-11e7-86de-6cc0f000f079.jpg)

It just makes it dimmer and worse to distinguish (at least for my eyes)
![screen shot 2017-06-12 at 13 06 52](https://user-images.githubusercontent.com/983242/27031295-07209270-4f70-11e7-9176-0db263217c67.jpg)
![screen shot 2017-06-12 at 13 06 56](https://user-images.githubusercontent.com/983242/27031296-07230730-4f70-11e7-8d4e-35bd5af794f7.jpg)

### Using 10 css weight classes instead of only 5
It increases a bit the color, but still scenario 1 its impossible to distinguish if you don't come very close to the screen.
![screen shot 2017-06-12 at 13 10 41](https://user-images.githubusercontent.com/983242/27031459-9f1f03e0-4f70-11e7-87c4-79983e21bc0b.jpg)
![screen shot 2017-06-12 at 13 10 45](https://user-images.githubusercontent.com/983242/27031458-9f17b0a4-4f70-11e7-953b-f29049a9e85f.jpg)

I've even tried to use 20 css weight classes but results are pretty similar... I think the main problem with scenario 1 is the fact that weights are so distant that the increase is impossible to see. When using 2+2 you're seen a double increase, but 3+1 its not perceptible.

**Comment by bertocq** [comment]
Well I've switched to [hsla](https://css-tricks.com/yay-for-hsla/) in order to make those Hue increments to go for all the "fire" range:

![screen shot 2017-06-12 at 13 28 10](https://user-images.githubusercontent.com/983242/27032032-264569b6-4f73-11e7-9568-0c51d3bbb8db.jpg)

And results let you distinguish clearly on both scenarios :)... but I've maybe gone to far into the design field ^_^ 

![screen shot 2017-06-12 at 13 30 08](https://user-images.githubusercontent.com/983242/27032046-43c10e46-4f73-11e7-923e-b643c6bda9a2.jpg)

I think its again time to check @decabeza opinion and possible suggestions

**Comment by voodoorai2000** [comment]
Looks reasonably awesome :) 

Let's see if we have @decabeza's blessing

**Comment by decabeza** [comment]
Good investigation job @bertocq! 

My first idea was try with different colours, but it's true looks some crazy 😝. 

Let me try some CSS magic tricks 🔮 so let's see if I can get something

**Comment by bertocq** [comment]
Just so I don't forget (and because it was on my git stash and I almost destroyed it) This is the css code of the last attempt that for me is the best looking among of all my tries :D
```css
.annotator-hl.weight-1 {
  background: hsla(60, 87%, 54%, 0.2);
}

.annotator-hl.weight-2 {
  background: hsla(45, 87%, 54%, 0.4);
}

.annotator-hl.weight-3 {
  background: hsla(30, 87%, 54%, 0.6);
}

.annotator-hl.weight-4 {
  background: hsla(15, 87%, 54%, 0.8);
}

.annotator-hl.weight-5 {
  background: hsla(0, 87%, 54%, 1);
}
```

![screen shot 2017-06-12 at 13 39 39](https://user-images.githubusercontent.com/983242/27032331-ab840424-4f74-11e7-877e-120909df6fbd.jpg)

**Comment by voodoorai2000** [comment]
Closed via #1711
